superpoint/models/utils.py

```diff
@@ -105,7 +106,18 @@ def descriptor_loss(descriptors, warped_descriptors, homographies, **config):
     positive_dist = tf.maximum(0., config['positive_margin'] - dot_product_desc)
     negative_dist = tf.maximum(0., dot_product_desc - config['negative_margin'])
     loss = config['lambda_d'] * s * positive_dist + (1 - s) * negative_dist
-    return tf.reduce_mean(loss)
+
+    # Mask the pixels if bordering artifacts appear
+    valid_mask = tf.ones([batch_size, Hc, Wc], tf.float32)\
+        if valid_mask is None else valid_mask
+    valid_mask = tf.to_float(valid_mask[..., tf.newaxis])  # for GPU
+    valid_mask = tf.space_to_depth(valid_mask, config['grid_size'])
+    valid_mask = tf.reduce_prod(valid_mask, axis=3)  # AND along the channel dim
+    valid_mask = tf.reshape(valid_mask, [batch_size, 1, 1, Hc, Wc])
+
+    normalization = tf.reduce_sum(valid_mask) * tf.reduce_sum(valid_mask) / batch_size
+    loss = tf.reduce_sum(valid_mask * loss) / normalization
```

Comment:
When broadcasting, the shape of `valid_mask` is changed from `[B, 1, 1, Hc, Wc]` to `[B, Hc, Wc, Hc, Wc]`. Thus shouldn't the normalization factor be `tf.reduce_sum(valid_mask) * Hc * Wc` instead ?

Reply:
Indeed, good point.